state/watcher.go

```diff
@@ -1105,39 +1127,58 @@ type lifeStatus struct {
 // of the specified relations change.
 type relationStatusWatcher struct {
 	commonWatcher
-	members         bson.D
-	filter          func(key interface{}) bool
-	knownLifeStatus map[string]lifeStatus
-	out             chan []string
+	lifeWatcher         StringsWatcher
+	knownLifeStatusById map[string]lifeStatus
+	out                 chan []string
+
+	// key filters are used to filter changes based on the
+	// relation key, ie life changes from the relation collection
+	relationKeyMatch  bson.D
+	relationKeyFilter func(key interface{}) bool
+
+	// status filter is used to filter changes based on the
+	// relation id number, ie status changes from the status collection
+	relationStatusFilter func(key interface{}) bool
 }
 
 // WatchStatus returns a watcher that notifies of changes to the life
 // or status of the relation.
```

Comment:
It needs to be in the one state watcher because we are watching 2 different collections, and need to maintain current state which covers life and status combined. Otherwise we would end up firing duplicate events when first life changes and then status (or visa versa).

Reply:
When you combine the two watchers, you wait for the initial events from both.

Reply:
It's a bit more complicated. The watcher can be set up from either a relation (watch only one entity) or an application (watch all relations for that application). The status collection is keyed on r#<id> not key. In the application case, we have a string match for the relation key, and so need to first query the relation collection to get the superset of things to watch and thus the set of ids. And when a new relation is created, we need that info in the combined watcher also to track the id. So things are quite tightly all bound up together because of the different ids used in each of the watched collections and how they are related.

Reply:
That's not how watchers are typically created. It would be less complicated (here, in state), if you went with the more standard approach:

 - Application.WatchRelations watches *lifecycle* changes, as described in the doc comment
 - For each active relation, watch that relation's status (Relation.WatchStatus *only* watches the status doc for the relation)

That's what I was trying to suggest earlier. This is what we do in various workers: we have a top-level lifecycle watcher which is notified of things being created or destroyed; we then start watchers for those individual entities that we care about (usually the Alive and Dying ones).

Is there a problem with the approach I describe?

Reply:
I was trying to implement a "bulk" watcher to minimise the number of workers as it seems to me our typical O(N) approach to doing things like this is a scalability issue and has always bothered me. My approach is more complex under the covers but scales much better. Nonetheless, I'll rework.

Reply:
state watchers reworked so that the WatchRelations() and WatchLifeStatus() apis construct a standard life watcher according to what they need - for each newly alive relation, a standard status watcher is started. Everything is managed by a new relationLifeStatus watcher.